// @cosmos-kit/react/main/modal/components/ConnectWalletCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleAvatarWithName = exports.ConnectWalletCard = void 0;
var _react = require("@chakra-ui/react");
var _react2 = _interopRequireDefault(require("react"));
var _io = require("react-icons/io");
var SimpleAvatarWithName = function SimpleAvatarWithName(_ref) {
  var walletIcon = _ref.walletIcon,
    username = _ref.username,
    icon = _ref.icon;
  return /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    spacing: 1,
    alignItems: "center"
  }, /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    display: icon ? 'block' : 'none',
    minW: 24,
    maxW: 24,
    w: 24,
    minH: 24,
    maxH: 24,
    h: 24,
    borderRadius: "full",
    overflow: "hidden"
  }, icon), username && /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    isInline: true,
    alignItems: "center"
  }, walletIcon ? /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    minW: 5,
    maxW: 5,
    w: 5,
    minH: 5,
    maxH: 5,
    h: 5
  }, /*#__PURE__*/_react2["default"].createElement(_react.Image, {
    src: walletIcon
  })) : /*#__PURE__*/_react2["default"].createElement(_react.Center, {
    borderRadius: "full",
    p: 1.5,
    bg: (0, _react.useColorModeValue)('purple.50', 'purple.700'),
    color: (0, _react.useColorModeValue)('purple.700', 'purple.100')
  }, /*#__PURE__*/_react2["default"].createElement(_react.Icon, {
    as: _io.IoMdWallet
  })), /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    fontSize: {
      md: 'xl'
    },
    fontWeight: "semibold"
  }, username)));
};
exports.SimpleAvatarWithName = SimpleAvatarWithName;
var ConnectWalletCard = function ConnectWalletCard(_ref2) {
  var userInfo = _ref2.userInfo,
    addressBtn = _ref2.addressBtn,
    connectWalletButton = _ref2.connectWalletButton;
  return /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    justifyContent: "center",
    alignItems: "center",
    borderRadius: "lg",
    bg: (0, _react.useColorModeValue)('white', 'blackAlpha.400'),
    boxShadow: (0, _react.useColorModeValue)('0 0 2px #dfdfdf, 0 0 6px -2px #d3d3d3', '0 0 2px #363636, 0 0 8px -2px #4f4f4f'),
    spacing: 4,
    px: 8,
    py: {
      base: 6,
      md: 12
    }
  }, userInfo, addressBtn, connectWalletButton);
};
exports.ConnectWalletCard = ConnectWalletCard;
