// @cosmos-kit/react/main/modal/components/SimpleConnectModalHead.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleModalHead = void 0;
var _react = require("@chakra-ui/react");
var _react2 = _interopRequireDefault(require("react"));
var _fi = require("react-icons/fi");
var _defaultComponent = require("./default-component");
var SimpleModalHead = function SimpleModalHead(_ref) {
  var title = _ref.title,
    backButton = _ref.backButton,
    handleBack = _ref.handleBack,
    handleClose = _ref.handleClose;
  var _useColorMode = (0, _react.useColorMode)(),
    colorMode = _useColorMode.colorMode;
  return /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    w: "full",
    isInline: true,
    alignItems: "center",
    h: "fit-content",
    mb: 1,
    p: 4,
    pb: 0
  }, backButton && /*#__PURE__*/_react2["default"].createElement(_react.Button, {
    variant: "ghost",
    borderRadius: "full",
    bg: "transparent",
    px: 0,
    _focus: {
      outline: 'none'
    },
    onClick: handleBack,
    color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'gray.800', 'white')
  }, /*#__PURE__*/_react2["default"].createElement(_react.Icon, {
    as: _fi.FiChevronLeft,
    w: 6,
    h: 6
  })), /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    flex: 1,
    mr: backButton ? 0 : -10
  }, /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    fontSize: "md",
    fontWeight: "semibold",
    textAlign: "center",
    color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'gray.800', 'white')
  }, title)), /*#__PURE__*/_react2["default"].createElement(_react.Button, {
    variant: "ghost",
    borderRadius: "full",
    bg: "transparent",
    px: 0,
    _focus: {
      outline: 'none'
    },
    onClick: handleClose,
    color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'gray.800', 'white')
  }, /*#__PURE__*/_react2["default"].createElement(_react.Icon, {
    as: _fi.FiX,
    w: 5,
    h: 5
  })));
};
exports.SimpleModalHead = SimpleModalHead;
