// @cosmos-kit/react/module/modal/components/SimpleConnectModalHeadV1.js
import { Box, Button, Icon, Stack, Text, useColorMode } from '@chakra-ui/react';
import React from 'react';
import { FiChevronLeft, FiX } from 'react-icons/fi';
import { handleChangeColorModeValue } from './default-component';
export const SimpleModalHeadV1 = ({
  title,
  backButton,
  handleBack,
  handleClose
}) => {
  const {
    colorMode
  } = useColorMode();
  return /*#__PURE__*/React.createElement(Stack, {
    w: "full",
    isInline: true,
    alignItems: "center",
    h: "fit-content",
    mb: 2.5,
    p: 4,
    pb: 0
  }, backButton && /*#__PURE__*/React.createElement(Button, {
    variant: "ghost",
    borderRadius: "full",
    px: 0,
    _focus: {
      outline: 'none'
    },
    onClick: handleBack,
    color: handleChangeColorModeValue(colorMode, 'gray.800', 'white')
  }, /*#__PURE__*/React.createElement(Icon, {
    as: FiChevronLeft,
    w: 6,
    h: 6
  })), /*#__PURE__*/React.createElement(Box, {
    flex: 1,
    pl: 2
  }, /*#__PURE__*/React.createElement(Text, {
    fontSize: "lg",
    fontWeight: "medium",
    textAlign: backButton ? 'center' : 'start',
    color: handleChangeColorModeValue(colorMode, 'gray.800', 'white')
  }, title)), /*#__PURE__*/React.createElement(Button, {
    variant: "ghost",
    borderRadius: "full",
    px: 0,
    _focus: {
      outline: 'none'
    },
    onClick: handleClose,
    color: handleChangeColorModeValue(colorMode, 'gray.800', 'white')
  }, /*#__PURE__*/React.createElement(Icon, {
    as: FiX,
    w: 5,
    h: 5
  })));
};
