// @cosmos-kit/core/main/bases/main-wallet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MainWalletBase = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _types = require("../types");
var _wallet = require("./wallet");
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var MainWalletBase = /*#__PURE__*/function (_WalletBase) {
  (0, _inherits2["default"])(MainWalletBase, _WalletBase);
  var _super = _createSuper(MainWalletBase);
  function MainWalletBase(walletInfo, ChainWallet) {
    var _this;
    (0, _classCallCheck2["default"])(this, MainWalletBase);
    _this = _super.call(this, walletInfo);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_chainWallets", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "preferredEndpoints", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "ChainWallet", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getChainWallet", function (chainName) {
      var _this$chainWallets;
      return (_this$chainWallets = _this.chainWallets) === null || _this$chainWallets === void 0 ? void 0 : _this$chainWallets.get(chainName);
    });
    _this.ChainWallet = ChainWallet;
    _this.clientPromise = _this.fetchClient();
    return _this;
  }
  (0, _createClass2["default"])(MainWalletBase, [{
    key: "onSetChainsDone",
    value: function onSetChainsDone() {
      var _this$chainWallets2,
        _this2 = this;
      (_this$chainWallets2 = this.chainWallets) === null || _this$chainWallets2 === void 0 ? void 0 : _this$chainWallets2.forEach(function (chainWallet) {
        chainWallet.client = _this2.client;
        chainWallet.clientPromise = _this2.clientPromise;
        chainWallet.fetchClient = _this2.fetchClient;
      });
    }
  }, {
    key: "setChains",
    value: function setChains(chains) {
      var _this3 = this;
      this._chainWallets = new Map(chains.map(function (chain) {
        var _chain$preferredEndpo, _this3$preferredEndpo, _this3$preferredEndpo2, _chain$chain, _chain$chain$apis, _chain$chain$apis$rpc, _chain$preferredEndpo2, _this3$preferredEndpo3, _this3$preferredEndpo4, _chain$chain2, _chain$chain2$apis, _chain$chain2$apis$re;
        chain.preferredEndpoints = {
          rpc: [].concat((0, _toConsumableArray2["default"])(((_chain$preferredEndpo = chain.preferredEndpoints) === null || _chain$preferredEndpo === void 0 ? void 0 : _chain$preferredEndpo.rpc) || []), (0, _toConsumableArray2["default"])(((_this3$preferredEndpo = _this3.preferredEndpoints) === null || _this3$preferredEndpo === void 0 ? void 0 : (_this3$preferredEndpo2 = _this3$preferredEndpo[chain.name]) === null || _this3$preferredEndpo2 === void 0 ? void 0 : _this3$preferredEndpo2.rpc) || []), ["https://rpc.cosmos.directory/".concat(chain.name)], (0, _toConsumableArray2["default"])(((_chain$chain = chain.chain) === null || _chain$chain === void 0 ? void 0 : (_chain$chain$apis = _chain$chain.apis) === null || _chain$chain$apis === void 0 ? void 0 : (_chain$chain$apis$rpc = _chain$chain$apis.rpc) === null || _chain$chain$apis$rpc === void 0 ? void 0 : _chain$chain$apis$rpc.map(function (e) {
            return e.address;
          })) || [])),
          rest: [].concat((0, _toConsumableArray2["default"])(((_chain$preferredEndpo2 = chain.preferredEndpoints) === null || _chain$preferredEndpo2 === void 0 ? void 0 : _chain$preferredEndpo2.rest) || []), (0, _toConsumableArray2["default"])(((_this3$preferredEndpo3 = _this3.preferredEndpoints) === null || _this3$preferredEndpo3 === void 0 ? void 0 : (_this3$preferredEndpo4 = _this3$preferredEndpo3[chain.name]) === null || _this3$preferredEndpo4 === void 0 ? void 0 : _this3$preferredEndpo4.rest) || []), ["https://rest.cosmos.directory/".concat(chain.name)], (0, _toConsumableArray2["default"])(((_chain$chain2 = chain.chain) === null || _chain$chain2 === void 0 ? void 0 : (_chain$chain2$apis = _chain$chain2.apis) === null || _chain$chain2$apis === void 0 ? void 0 : (_chain$chain2$apis$re = _chain$chain2$apis.rest) === null || _chain$chain2$apis$re === void 0 ? void 0 : _chain$chain2$apis$re.map(function (e) {
            return e.address;
          })) || []))
        };
        return [chain.name, new _this3.ChainWallet(_this3.walletInfo, chain)];
      }));
      this.onSetChainsDone();
    }
  }, {
    key: "username",
    get: function get() {
      var _this$data;
      return (_this$data = this.data) === null || _this$data === void 0 ? void 0 : _this$data.username;
    }
  }, {
    key: "chainWallets",
    get: function get() {
      return this._chainWallets;
    }
  }, {
    key: "update",
    value: function () {
      var _update = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(sessionOptions, callbacks) {
        var _ref,
          _ref$beforeConnect,
          _this4 = this,
          _ref2,
          _ref2$afterConnect;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return (_ref = callbacks || this.callbacks) === null || _ref === void 0 ? void 0 : (_ref$beforeConnect = _ref.beforeConnect) === null || _ref$beforeConnect === void 0 ? void 0 : _ref$beforeConnect.call(_ref);
              case 2:
                if (this.client) {
                  _context.next = 5;
                  break;
                }
                this.setClientNotExist();
                return _context.abrupt("return");
              case 5:
                this.setState(_types.State.Done);
                if (sessionOptions !== null && sessionOptions !== void 0 && sessionOptions.duration) {
                  setTimeout(function () {
                    _this4.disconnect(callbacks);
                  }, sessionOptions === null || sessionOptions === void 0 ? void 0 : sessionOptions.duration);
                }
                _context.next = 9;
                return (_ref2 = callbacks || this.callbacks) === null || _ref2 === void 0 ? void 0 : (_ref2$afterConnect = _ref2.afterConnect) === null || _ref2$afterConnect === void 0 ? void 0 : _ref2$afterConnect.call(_ref2);
              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
      function update(_x, _x2) {
        return _update.apply(this, arguments);
      }
      return update;
    }()
  }, {
    key: "reset",
    value: function reset() {
      var _this$chainWallets3;
      (_this$chainWallets3 = this.chainWallets) === null || _this$chainWallets3 === void 0 ? void 0 : _this$chainWallets3.forEach(function (chain) {
        chain.reset();
      });
      this.setData(undefined);
      this.setMessage(undefined);
      this.setState(_types.State.Init);
    }
  }]);
  return MainWalletBase;
}(_wallet.WalletBase);
exports.MainWalletBase = MainWalletBase;
