// @cosmos-kit/core/main/manager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WalletManager = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _bowser = _interopRequireDefault(require("bowser"));
var _bases = require("./bases");
var _types = require("./types");
var _utils = require("./utils");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var WalletManager = /*#__PURE__*/function (_StateBase) {
  (0, _inherits2["default"])(WalletManager, _StateBase);
  var _super = _createSuper(WalletManager);
  function WalletManager(_chains, _assetLists, _wallets, _signerOptions, viewOptions, _endpointOptions, storageOptions, sessionOptions) {
    var _this;
    (0, _classCallCheck2["default"])(this, WalletManager);
    _this = _super.call(this);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_currentWalletName", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_currentChainName", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_activeWallets", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_totalWallets", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_chainRecords", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "viewOptions", {
      alwaysOpenView: false,
      closeViewWhenWalletIsConnected: false,
      closeViewWhenWalletIsDisconnected: true,
      closeViewWhenWalletIsRejected: false
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "storageOptions", {
      disabled: false,
      duration: 1800000,
      clearOnTabClose: false
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "sessionOptions", {
      duration: 1800000,
      killOnTabClose: false
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setChains", function (chains, assetLists, signerOptions, endpointOptions) {
      _this._chainRecords = chains.map(function (chain) {
        return (0, _utils.convertChain)(chain, assetLists, signerOptions, endpointOptions === null || endpointOptions === void 0 ? void 0 : endpointOptions[chain.chain_name]);
      });
      console.info("".concat(_this.chainCount, " chains are available!"));
      _this._totalWallets.forEach(function (wallet) {
        wallet.setChains(_this._chainRecords);
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setWallets", function (wallets) {
      _this._totalWallets = wallets;
      console.info("".concat(_this.walletCount, " wallets are available!"));
      _this._totalWallets.forEach(function (wallet) {
        wallet.setChains(_this._chainRecords);
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setActiveWalletNames", function (walletNames) {
      if (walletNames) {
        _this._activeWallets = _this._totalWallets.filter(function (wallet) {
          return walletNames.includes(wallet.walletName);
        });
      } else {
        _this._activeWallets = _this._totalWallets;
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "enable", /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(chainIds) {
        var _this$currentWallet, _this$currentWallet$c, _this$currentWallet$c2;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return (_this$currentWallet = _this.currentWallet) === null || _this$currentWallet === void 0 ? void 0 : (_this$currentWallet$c = _this$currentWallet.client) === null || _this$currentWallet$c === void 0 ? void 0 : (_this$currentWallet$c2 = _this$currentWallet$c.enable) === null || _this$currentWallet$c2 === void 0 ? void 0 : _this$currentWallet$c2.call(_this$currentWallet$c, chainIds);
              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getRpcEndpoint", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
      var _this$currentWallet2, _this$currentWallet2$;
      return _regenerator["default"].wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return (_this$currentWallet2 = _this.currentWallet) === null || _this$currentWallet2 === void 0 ? void 0 : (_this$currentWallet2$ = _this$currentWallet2.getRpcEndpoint) === null || _this$currentWallet2$ === void 0 ? void 0 : _this$currentWallet2$.call(_this$currentWallet2);
            case 2:
              return _context2.abrupt("return", _context2.sent);
            case 3:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getRestEndpoint", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
      var _this$currentWallet3, _this$currentWallet3$;
      return _regenerator["default"].wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return (_this$currentWallet3 = _this.currentWallet) === null || _this$currentWallet3 === void 0 ? void 0 : (_this$currentWallet3$ = _this$currentWallet3.getRestEndpoint) === null || _this$currentWallet3$ === void 0 ? void 0 : _this$currentWallet3$.call(_this$currentWallet3);
            case 2:
              return _context3.abrupt("return", _context3.sent);
            case 3:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getStargateClient", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
      var _this$currentWallet4, _this$currentWallet4$;
      return _regenerator["default"].wrap(function _callee4$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return (_this$currentWallet4 = _this.currentWallet) === null || _this$currentWallet4 === void 0 ? void 0 : (_this$currentWallet4$ = _this$currentWallet4.getStargateClient) === null || _this$currentWallet4$ === void 0 ? void 0 : _this$currentWallet4$.call(_this$currentWallet4);
            case 2:
              return _context4.abrupt("return", _context4.sent);
            case 3:
            case "end":
              return _context4.stop();
          }
        }
      }, _callee4);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getCosmWasmClient", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
      var _this$currentWallet5, _this$currentWallet5$;
      return _regenerator["default"].wrap(function _callee5$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              _context5.next = 2;
              return (_this$currentWallet5 = _this.currentWallet) === null || _this$currentWallet5 === void 0 ? void 0 : (_this$currentWallet5$ = _this$currentWallet5.getCosmWasmClient) === null || _this$currentWallet5$ === void 0 ? void 0 : _this$currentWallet5$.call(_this$currentWallet5);
            case 2:
              return _context5.abrupt("return", _context5.sent);
            case 3:
            case "end":
              return _context5.stop();
          }
        }
      }, _callee5);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getSigningStargateClient", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
      var _this$currentWallet6, _this$currentWallet6$;
      return _regenerator["default"].wrap(function _callee6$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return (_this$currentWallet6 = _this.currentWallet) === null || _this$currentWallet6 === void 0 ? void 0 : (_this$currentWallet6$ = _this$currentWallet6.getSigningStargateClient) === null || _this$currentWallet6$ === void 0 ? void 0 : _this$currentWallet6$.call(_this$currentWallet6);
            case 2:
              return _context6.abrupt("return", _context6.sent);
            case 3:
            case "end":
              return _context6.stop();
          }
        }
      }, _callee6);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getSigningCosmWasmClient", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7() {
      var _this$currentWallet7, _this$currentWallet7$;
      return _regenerator["default"].wrap(function _callee7$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              _context7.next = 2;
              return (_this$currentWallet7 = _this.currentWallet) === null || _this$currentWallet7 === void 0 ? void 0 : (_this$currentWallet7$ = _this$currentWallet7.getSigningCosmWasmClient) === null || _this$currentWallet7$ === void 0 ? void 0 : _this$currentWallet7$.call(_this$currentWallet7);
            case 2:
              return _context7.abrupt("return", _context7.sent);
            case 3:
            case "end":
              return _context7.stop();
          }
        }
      }, _callee7);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "sign", /*#__PURE__*/function () {
      var _ref8 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(messages, fee, memo, type) {
        var _this$currentWallet8, _this$currentWallet8$;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return (_this$currentWallet8 = _this.currentWallet) === null || _this$currentWallet8 === void 0 ? void 0 : (_this$currentWallet8$ = _this$currentWallet8.sign) === null || _this$currentWallet8$ === void 0 ? void 0 : _this$currentWallet8$.call(_this$currentWallet8, messages, fee, memo, type);
              case 2:
                return _context8.abrupt("return", _context8.sent);
              case 3:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8);
      }));
      return function (_x2, _x3, _x4, _x5) {
        return _ref8.apply(this, arguments);
      };
    }());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "broadcast", /*#__PURE__*/function () {
      var _ref9 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(signedMessages, type) {
        var _this$currentWallet9, _this$currentWallet9$;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return (_this$currentWallet9 = _this.currentWallet) === null || _this$currentWallet9 === void 0 ? void 0 : (_this$currentWallet9$ = _this$currentWallet9.broadcast) === null || _this$currentWallet9$ === void 0 ? void 0 : _this$currentWallet9$.call(_this$currentWallet9, signedMessages, type);
              case 2:
                return _context9.abrupt("return", _context9.sent);
              case 3:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9);
      }));
      return function (_x6, _x7) {
        return _ref9.apply(this, arguments);
      };
    }());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "signAndBroadcast", /*#__PURE__*/function () {
      var _ref10 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(messages, fee, memo, type) {
        var _this$currentWallet10, _this$currentWallet11;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return (_this$currentWallet10 = _this.currentWallet) === null || _this$currentWallet10 === void 0 ? void 0 : (_this$currentWallet11 = _this$currentWallet10.signAndBroadcast) === null || _this$currentWallet11 === void 0 ? void 0 : _this$currentWallet11.call(_this$currentWallet10, messages, fee, memo, type);
              case 2:
                return _context10.abrupt("return", _context10.sent);
              case 3:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10);
      }));
      return function (_x8, _x9, _x10, _x11) {
        return _ref10.apply(this, arguments);
      };
    }());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "reset", function () {
      var _this$currentWallet12;
      (_this$currentWallet12 = _this.currentWallet) === null || _this$currentWallet12 === void 0 ? void 0 : _this$currentWallet12.reset();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "updateLocalStorage", function (target) {
      var _window, _window2;
      if (!_this.useStorage) {
        return;
      }
      var storeObj = {};
      var storeStr = (_window = window) === null || _window === void 0 ? void 0 : _window.localStorage.getItem('cosmos-kit');
      if (storeStr) {
        storeObj = JSON.parse(storeStr);
      }
      switch (target) {
        case 'chain':
          storeObj = _objectSpread(_objectSpread({}, storeObj), {}, {
            currentChainName: _this.currentChainName
          });
          break;
        case 'wallet':
          storeObj = _objectSpread(_objectSpread({}, storeObj), {}, {
            currentWalletName: _this.currentWalletName
          });
          break;
        default:
          storeObj = _objectSpread(_objectSpread({}, storeObj), {}, {
            currentWalletName: _this.currentWalletName,
            currentChainName: _this.currentChainName
          });
          break;
      }
      (_window2 = window) === null || _window2 === void 0 ? void 0 : _window2.localStorage.setItem('cosmos-kit', JSON.stringify(storeObj));
      if (_this.storageOptions.duration) {
        setTimeout(function () {
          var _window3;
          (_window3 = window) === null || _window3 === void 0 ? void 0 : _window3.localStorage.removeItem('cosmos-kit');
        }, _this.storageOptions.duration);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setCurrentWallet", function (walletName) {
      var _this$emitWalletName, _this2;
      _this.reset();
      _this._currentWalletName = walletName;
      (_this$emitWalletName = (_this2 = _this).emitWalletName) === null || _this$emitWalletName === void 0 ? void 0 : _this$emitWalletName.call(_this2, walletName);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setCurrentChain", function (chainName) {
      var _this$emitChainName, _this3;
      _this.reset();
      _this._currentChainName = chainName;
      (_this$emitChainName = (_this3 = _this).emitChainName) === null || _this$emitChainName === void 0 ? void 0 : _this$emitChainName.call(_this3, chainName);
      _this.updateLocalStorage('chain');
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getWallet", function (walletName, chainName) {
      var wallet = _this._totalWallets.find(function (w) {
        return w.walletName === walletName;
      });
      if (!wallet) {
        throw new Error("".concat(walletName, " is not provided!"));
      }
      if (chainName) {
        wallet = wallet.getChainWallet(chainName);
        if (!wallet) {
          throw new Error("Unknown chain name: ".concat(chainName));
        }
      }
      wallet.actions = _this.actions;
      return wallet;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getChainRecord", function (chainName) {
      if (!chainName) {
        return void 0;
      }
      var chainRecord = _this.chainRecords.find(function (c) {
        return c.name === chainName;
      });
      if (!chainRecord) {
        throw new Error("".concat(chainName, " is not provided!"));
      }
      return chainRecord;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getChainLogo", function (chainName) {
      var _chainRecord$assetLis, _chainRecord$assetLis2, _chainRecord$assetLis3, _chainRecord$assetLis4, _chainRecord$assetLis5, _chainRecord$assetLis6;
      var chainRecord = _this.getChainRecord(chainName);
      return (
        // until chain_registry fix this
        // chainRecord?.chain.logo_URIs?.svg ||
        // chainRecord?.chain.logo_URIs?.png ||
        // chainRecord?.chain.logo_URIs?.jpeg ||
        (chainRecord === null || chainRecord === void 0 ? void 0 : (_chainRecord$assetLis = chainRecord.assetList) === null || _chainRecord$assetLis === void 0 ? void 0 : (_chainRecord$assetLis2 = _chainRecord$assetLis.assets[0]) === null || _chainRecord$assetLis2 === void 0 ? void 0 : (_chainRecord$assetLis3 = _chainRecord$assetLis2.logo_URIs) === null || _chainRecord$assetLis3 === void 0 ? void 0 : _chainRecord$assetLis3.svg) || (chainRecord === null || chainRecord === void 0 ? void 0 : (_chainRecord$assetLis4 = chainRecord.assetList) === null || _chainRecord$assetLis4 === void 0 ? void 0 : (_chainRecord$assetLis5 = _chainRecord$assetLis4.assets[0]) === null || _chainRecord$assetLis5 === void 0 ? void 0 : (_chainRecord$assetLis6 = _chainRecord$assetLis5.logo_URIs) === null || _chainRecord$assetLis6 === void 0 ? void 0 : _chainRecord$assetLis6.png) || undefined
      );
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "connect", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11() {
      var _this$viewOptions;
      var current, _this$viewOptions2, _this$viewOptions3;
      return _regenerator["default"].wrap(function _callee11$(_context11) {
        while (1) {
          switch (_context11.prev = _context11.next) {
            case 0:
              current = _this.currentWallet;
              if (current) {
                _context11.next = 4;
                break;
              }
              _this.openView();
              return _context11.abrupt("return");
            case 4:
              if ((_this$viewOptions = _this.viewOptions) !== null && _this$viewOptions !== void 0 && _this$viewOptions.alwaysOpenView) {
                _this.openView();
              }
              if (current.isWalletNotExist || current.isWalletError) {
                _this.openView();
              }
              _context11.prev = 6;
              current.setEnv(_this.env);
              _context11.next = 10;
              return current.connect(_this.sessionOptions, _this.callbacks);
            case 10:
              if (_this.isWalletConnected && (_this$viewOptions2 = _this.viewOptions) !== null && _this$viewOptions2 !== void 0 && _this$viewOptions2.closeViewWhenWalletIsConnected) {
                _this.closeView();
              }
              _context11.next = 17;
              break;
            case 13:
              _context11.prev = 13;
              _context11.t0 = _context11["catch"](6);
              console.error(_context11.t0);
              if (_this.isWalletRejected && (_this$viewOptions3 = _this.viewOptions) !== null && _this$viewOptions3 !== void 0 && _this$viewOptions3.closeViewWhenWalletIsRejected) {
                _this.closeView();
              }
            case 17:
            case "end":
              return _context11.stop();
          }
        }
      }, _callee11, null, [[6, 13]]);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "disconnect", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12() {
      var _this$viewOptions4;
      var current, _this$viewOptions5, _this$viewOptions6;
      return _regenerator["default"].wrap(function _callee12$(_context12) {
        while (1) {
          switch (_context12.prev = _context12.next) {
            case 0:
              current = _this.currentWallet;
              if (current) {
                _context12.next = 4;
                break;
              }
              _this.setMessage('Current Wallet not defined.');
              return _context12.abrupt("return");
            case 4:
              if ((_this$viewOptions4 = _this.viewOptions) !== null && _this$viewOptions4 !== void 0 && _this$viewOptions4.alwaysOpenView) {
                _this.openView();
              }
              _context12.prev = 5;
              _context12.next = 8;
              return current.disconnect(_this.callbacks);
            case 8:
              if (_this.isWalletConnected && (_this$viewOptions5 = _this.viewOptions) !== null && _this$viewOptions5 !== void 0 && _this$viewOptions5.closeViewWhenWalletIsDisconnected) {
                _this.closeView();
              }
              _context12.next = 15;
              break;
            case 11:
              _context12.prev = 11;
              _context12.t0 = _context12["catch"](5);
              _this.setMessage(_context12.t0.message);
              if (_this.isWalletRejected && (_this$viewOptions6 = _this.viewOptions) !== null && _this$viewOptions6 !== void 0 && _this$viewOptions6.closeViewWhenWalletIsRejected) {
                _this.closeView();
              }
            case 15:
            case "end":
              return _context12.stop();
          }
        }
      }, _callee12, null, [[5, 11]]);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "openView", function () {
      var _this$emitViewOpen, _this4;
      (_this$emitViewOpen = (_this4 = _this).emitViewOpen) === null || _this$emitViewOpen === void 0 ? void 0 : _this$emitViewOpen.call(_this4, true);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "closeView", function () {
      var _this$emitViewOpen2, _this5;
      (_this$emitViewOpen2 = (_this5 = _this).emitViewOpen) === null || _this$emitViewOpen2 === void 0 ? void 0 : _this$emitViewOpen2.call(_this5, false);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleTabLoad", function (event) {
      event === null || event === void 0 ? void 0 : event.preventDefault();
      _this.connect();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleTabClose", function (event) {
      event.preventDefault();
      if (_this.storageOptions.clearOnTabClose) {
        window.localStorage.removeItem('cosmos-kit');
      }
      if (_this.sessionOptions.killOnTabClose || _this.isWalletConnecting) {
        _this.disconnect();
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_connectEventListener", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13() {
      return _regenerator["default"].wrap(function _callee13$(_context13) {
        while (1) {
          switch (_context13.prev = _context13.next) {
            case 0:
              if (_this.isInit) {
                _context13.next = 3;
                break;
              }
              _context13.next = 3;
              return _this.connect();
            case 3:
            case "end":
              return _context13.stop();
          }
        }
      }, _callee13);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMounted", function () {
      if (typeof window === 'undefined') {
        return;
      }
      var parser = _bowser["default"].getParser(window.navigator.userAgent);
      _this.setEnv({
        browser: parser.getBrowserName(true),
        device: parser.getPlatform().type || 'desktop',
        os: parser.getOSName(true)
      });
      if (_this.useStorage) {
        var storeStr = window.localStorage.getItem('cosmos-kit');
        if (storeStr) {
          var _JSON$parse = JSON.parse(storeStr),
            currentWalletName = _JSON$parse.currentWalletName,
            currentChainName = _JSON$parse.currentChainName;
          _this.setCurrentWallet(currentWalletName);
          _this.setCurrentChain(currentChainName);
          if (currentWalletName) {
            if (document.readyState !== 'complete') {
              window.addEventListener('load', _this._handleTabLoad);
            } else {
              _this._handleTabLoad();
            }
          }
        }
        window.addEventListener('beforeunload', _this._handleTabClose);
        _this.wallets.forEach(function (wallet) {
          var _wallet$walletInfo$co, _wallet$walletInfo$co2;
          (_wallet$walletInfo$co = wallet.walletInfo.connectEventNamesOnWindow) === null || _wallet$walletInfo$co === void 0 ? void 0 : _wallet$walletInfo$co.forEach(function (eventName) {
            window.addEventListener(eventName, _this._connectEventListener);
          });
          (_wallet$walletInfo$co2 = wallet.walletInfo.connectEventNamesOnClient) === null || _wallet$walletInfo$co2 === void 0 ? void 0 : _wallet$walletInfo$co2.forEach( /*#__PURE__*/function () {
            var _ref14 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14(eventName) {
              var _ref15, _ref15$on;
              return _regenerator["default"].wrap(function _callee14$(_context14) {
                while (1) {
                  switch (_context14.prev = _context14.next) {
                    case 0:
                      _context14.t1 = wallet.client;
                      if (_context14.t1) {
                        _context14.next = 5;
                        break;
                      }
                      _context14.next = 4;
                      return wallet.clientPromise;
                    case 4:
                      _context14.t1 = _context14.sent;
                    case 5:
                      _context14.t2 = _ref15 = _context14.t1;
                      _context14.t0 = _context14.t2 === null;
                      if (_context14.t0) {
                        _context14.next = 9;
                        break;
                      }
                      _context14.t0 = _ref15 === void 0;
                    case 9:
                      if (!_context14.t0) {
                        _context14.next = 13;
                        break;
                      }
                      void 0;
                      _context14.next = 14;
                      break;
                    case 13:
                      (_ref15$on = _ref15.on) === null || _ref15$on === void 0 ? void 0 : _ref15$on.call(_ref15, eventName, _this._connectEventListener);
                    case 14:
                    case "end":
                      return _context14.stop();
                  }
                }
              }, _callee14);
            }));
            return function (_x12) {
              return _ref14.apply(this, arguments);
            };
          }());
        });
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onUnmounted", function () {
      if (typeof window === 'undefined') {
        return;
      }
      window.removeEventListener('beforeunload', _this._handleTabClose);
      window.removeEventListener('load', _this._handleTabLoad);
      _this.wallets.forEach(function (wallet) {
        var _wallet$walletInfo$co3, _wallet$walletInfo$co4;
        (_wallet$walletInfo$co3 = wallet.walletInfo.connectEventNamesOnWindow) === null || _wallet$walletInfo$co3 === void 0 ? void 0 : _wallet$walletInfo$co3.forEach(function (eventName) {
          window.removeEventListener(eventName, _this._connectEventListener);
        });
        (_wallet$walletInfo$co4 = wallet.walletInfo.connectEventNamesOnClient) === null || _wallet$walletInfo$co4 === void 0 ? void 0 : _wallet$walletInfo$co4.forEach( /*#__PURE__*/function () {
          var _ref16 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee15(eventName) {
            var _ref17, _ref17$off;
            return _regenerator["default"].wrap(function _callee15$(_context15) {
              while (1) {
                switch (_context15.prev = _context15.next) {
                  case 0:
                    _context15.t1 = wallet.client;
                    if (_context15.t1) {
                      _context15.next = 5;
                      break;
                    }
                    _context15.next = 4;
                    return wallet.clientPromise;
                  case 4:
                    _context15.t1 = _context15.sent;
                  case 5:
                    _context15.t2 = _ref17 = _context15.t1;
                    _context15.t0 = _context15.t2 === null;
                    if (_context15.t0) {
                      _context15.next = 9;
                      break;
                    }
                    _context15.t0 = _ref17 === void 0;
                  case 9:
                    if (!_context15.t0) {
                      _context15.next = 13;
                      break;
                    }
                    void 0;
                    _context15.next = 14;
                    break;
                  case 13:
                    (_ref17$off = _ref17.off) === null || _ref17$off === void 0 ? void 0 : _ref17$off.call(_ref17, eventName, _this._connectEventListener);
                  case 14:
                  case "end":
                    return _context15.stop();
                }
              }
            }, _callee15);
          }));
          return function (_x13) {
            return _ref16.apply(this, arguments);
          };
        }());
      });
    });
    _this.setWallets(_wallets);
    _this.setActiveWalletNames();
    _this.setChains(_chains, _assetLists, _signerOptions, _endpointOptions);
    _this.viewOptions = _objectSpread(_objectSpread({}, _this.viewOptions), viewOptions);
    _this.storageOptions = _objectSpread(_objectSpread({}, _this.storageOptions), storageOptions);
    _this.sessionOptions = _objectSpread(_objectSpread({}, _this.sessionOptions), sessionOptions);
    return _this;
  }
  (0, _createClass2["default"])(WalletManager, [{
    key: "wallets",
    get: function get() {
      if (this.isMobile) {
        return this._activeWallets.filter(function (wallet) {
          return !wallet.walletInfo.mobileDisabled;
        });
      }
      return this._activeWallets;
    }
  }, {
    key: "chainRecords",
    get: function get() {
      return this._chainRecords;
    }
  }, {
    key: "useStorage",
    get: function get() {
      return !this.storageOptions.disabled;
    }
  }, {
    key: "currentWalletName",
    get: function get() {
      if (!this._currentWalletName && this.walletCount === 1) {
        return this._totalWallets[0].walletName;
      }
      return this._currentWalletName;
    }
  }, {
    key: "currentChainName",
    get: function get() {
      if (!this._currentChainName && this.chainCount === 1) {
        return this.chainRecords[0].name;
      }
      return this._currentChainName;
    }
  }, {
    key: "currentWallet",
    get: function get() {
      return this.currentWalletName ? this.getWallet(this.currentWalletName, this.currentChainName) : void 0;
    }
  }, {
    key: "currentWalletInfo",
    get: function get() {
      var _this$currentWallet13;
      return (_this$currentWallet13 = this.currentWallet) === null || _this$currentWallet13 === void 0 ? void 0 : _this$currentWallet13.walletInfo;
    }
  }, {
    key: "currentChainRecord",
    get: function get() {
      return this.getChainRecord(this.currentChainName);
    }
  }, {
    key: "data",
    get: function get() {
      var _this$currentWallet14;
      return (_this$currentWallet14 = this.currentWallet) === null || _this$currentWallet14 === void 0 ? void 0 : _this$currentWallet14.data;
    }
  }, {
    key: "state",
    get: function get() {
      var _this$currentWallet15;
      return ((_this$currentWallet15 = this.currentWallet) === null || _this$currentWallet15 === void 0 ? void 0 : _this$currentWallet15.state) || _types.State.Init;
    }
  }, {
    key: "message",
    get: function get() {
      var _this$currentWallet16;
      return (_this$currentWallet16 = this.currentWallet) === null || _this$currentWallet16 === void 0 ? void 0 : _this$currentWallet16.message;
    }
  }, {
    key: "username",
    get: function get() {
      var _this$data;
      return (_this$data = this.data) === null || _this$data === void 0 ? void 0 : _this$data.username;
    }
  }, {
    key: "address",
    get: function get() {
      var _this$data2;
      return (_this$data2 = this.data) === null || _this$data2 === void 0 ? void 0 : _this$data2.address;
    }
  }, {
    key: "offlineSigner",
    get: function get() {
      var _this$data3;
      return (_this$data3 = this.data) === null || _this$data3 === void 0 ? void 0 : _this$data3.offlineSigner;
    }
  }, {
    key: "walletNames",
    get: function get() {
      return this._totalWallets.map(function (wallet) {
        return wallet.walletName;
      });
    }
  }, {
    key: "walletCount",
    get: function get() {
      return this.walletNames.length;
    }
  }, {
    key: "chainNames",
    get: function get() {
      return this.chainRecords.map(function (chain) {
        return chain.name;
      });
    }
  }, {
    key: "chainCount",
    get: function get() {
      return this.chainNames.length;
    }
  }, {
    key: "emitWalletName",
    get: function get() {
      var _this$actions;
      return (_this$actions = this.actions) === null || _this$actions === void 0 ? void 0 : _this$actions.walletName;
    }
  }, {
    key: "emitChainName",
    get: function get() {
      var _this$actions2;
      return (_this$actions2 = this.actions) === null || _this$actions2 === void 0 ? void 0 : _this$actions2.chainName;
    }
  }, {
    key: "emitViewOpen",
    get: function get() {
      var _this$actions3;
      return (_this$actions3 = this.actions) === null || _this$actions3 === void 0 ? void 0 : _this$actions3.viewOpen;
    }
  }, {
    key: "callbacks",
    get: function get() {
      var _this6 = this;
      return {
        afterConnect: function afterConnect() {
          if (!_this6.isWalletDisconnected) {
            _this6.updateLocalStorage('wallet');
          }
        },
        afterDisconnect: function afterDisconnect() {
          _this6.setCurrentWallet(undefined);
          _this6.updateLocalStorage('wallet');
        }
      };
    }
  }]);
  return WalletManager;
}(_bases.StateBase);
exports.WalletManager = WalletManager;
