// @cosmos-kit/core/main/manager-v2.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WalletManagerV2 = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _bowser = _interopRequireDefault(require("bowser"));
var _bases = require("./bases");
var _repository = require("./repository");
var _utils = require("./utils");
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var WalletManagerV2 = /*#__PURE__*/function (_StateBase) {
  (0, _inherits2["default"])(WalletManagerV2, _StateBase);
  var _super = _createSuper(WalletManagerV2);
  function WalletManagerV2(chains, assetLists, wallets, signerOptions, endpointOptions, sessionOptions) {
    var _this;
    (0, _classCallCheck2["default"])(this, WalletManagerV2);
    _this = _super.call(this);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "chainRecords", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "walletRepos", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "options", {
      synchroMutexWallet: true
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "sessionOptions", {
      duration: 1800000,
      killOnTabClose: false
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getWalletRepo", function (chainName) {
      var walletRepo = _this.walletRepos.find(function (wr) {
        return wr.chainName === chainName;
      });
      if (!walletRepo) {
        throw new Error("Chain ".concat(chainName, " is not provided."));
      }
      return walletRepo;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getChainRecord", function (chainName) {
      if (!chainName) {
        return void 0;
      }
      var chainRecord = _this.chainRecords.find(function (c) {
        return c.name === chainName;
      });
      if (!chainRecord) {
        throw new Error("".concat(chainName, " is not provided!"));
      }
      return chainRecord;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getChainLogo", function (chainName) {
      var _chainRecord$assetLis, _chainRecord$assetLis2, _chainRecord$assetLis3, _chainRecord$assetLis4, _chainRecord$assetLis5, _chainRecord$assetLis6;
      var chainRecord = _this.getChainRecord(chainName);
      return (
        // until chain_registry fix this
        // chainRecord?.chain.logo_URIs?.svg ||
        // chainRecord?.chain.logo_URIs?.png ||
        // chainRecord?.chain.logo_URIs?.jpeg ||
        (chainRecord === null || chainRecord === void 0 ? void 0 : (_chainRecord$assetLis = chainRecord.assetList) === null || _chainRecord$assetLis === void 0 ? void 0 : (_chainRecord$assetLis2 = _chainRecord$assetLis.assets[0]) === null || _chainRecord$assetLis2 === void 0 ? void 0 : (_chainRecord$assetLis3 = _chainRecord$assetLis2.logo_URIs) === null || _chainRecord$assetLis3 === void 0 ? void 0 : _chainRecord$assetLis3.svg) || (chainRecord === null || chainRecord === void 0 ? void 0 : (_chainRecord$assetLis4 = chainRecord.assetList) === null || _chainRecord$assetLis4 === void 0 ? void 0 : (_chainRecord$assetLis5 = _chainRecord$assetLis4.assets[0]) === null || _chainRecord$assetLis5 === void 0 ? void 0 : (_chainRecord$assetLis6 = _chainRecord$assetLis5.logo_URIs) === null || _chainRecord$assetLis6 === void 0 ? void 0 : _chainRecord$assetLis6.png) || undefined
      );
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_handleConnect", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
      var ls, walletName, _this$walletReposInUs;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              ls = window.localStorage;
              walletName = ls.getItem('chain-provider');
              if (!walletName) {
                _context.next = 6;
                break;
              }
              ls.setItem('synchronize-mutex-wallet', 'fire');
              _context.next = 6;
              return (_this$walletReposInUs = _this.walletReposInUse[0]) === null || _this$walletReposInUs === void 0 ? void 0 : _this$walletReposInUs.connect(walletName);
            case 6:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMounted", function () {
      var _this$walletRepos$;
      if (typeof window === 'undefined') {
        return;
      }
      _this._handleConnect();
      var parser = _bowser["default"].getParser(window.navigator.userAgent);
      var env = {
        browser: parser.getBrowserName(true),
        device: parser.getPlatform().type || 'desktop',
        os: parser.getOSName(true)
      };
      _this.setEnv(env);
      _this.walletRepos.forEach(function (repo) {
        return repo.setEnv(env);
      });
      (_this$walletRepos$ = _this.walletRepos[0]) === null || _this$walletRepos$ === void 0 ? void 0 : _this$walletRepos$.wallets.forEach(function (wallet) {
        var _wallet$walletInfo$co, _wallet$walletInfo$co2;
        (_wallet$walletInfo$co = wallet.walletInfo.connectEventNamesOnWindow) === null || _wallet$walletInfo$co === void 0 ? void 0 : _wallet$walletInfo$co.forEach(function (eventName) {
          window.addEventListener(eventName, _this._handleConnect);
        });
        (_wallet$walletInfo$co2 = wallet.walletInfo.connectEventNamesOnClient) === null || _wallet$walletInfo$co2 === void 0 ? void 0 : _wallet$walletInfo$co2.forEach( /*#__PURE__*/function () {
          var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(eventName) {
            var _ref3, _ref3$on;
            return _regenerator["default"].wrap(function _callee2$(_context2) {
              while (1) {
                switch (_context2.prev = _context2.next) {
                  case 0:
                    _context2.t1 = wallet.client;
                    if (_context2.t1) {
                      _context2.next = 5;
                      break;
                    }
                    _context2.next = 4;
                    return wallet.clientPromise;
                  case 4:
                    _context2.t1 = _context2.sent;
                  case 5:
                    _context2.t2 = _ref3 = _context2.t1;
                    _context2.t0 = _context2.t2 === null;
                    if (_context2.t0) {
                      _context2.next = 9;
                      break;
                    }
                    _context2.t0 = _ref3 === void 0;
                  case 9:
                    if (!_context2.t0) {
                      _context2.next = 13;
                      break;
                    }
                    void 0;
                    _context2.next = 14;
                    break;
                  case 13:
                    (_ref3$on = _ref3.on) === null || _ref3$on === void 0 ? void 0 : _ref3$on.call(_ref3, eventName, _this._handleConnect);
                  case 14:
                  case "end":
                    return _context2.stop();
                }
              }
            }, _callee2);
          }));
          return function (_x) {
            return _ref2.apply(this, arguments);
          };
        }());
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onUnmounted", function () {
      var _this$walletRepos$2;
      if (typeof window === 'undefined') {
        return;
      }
      (_this$walletRepos$2 = _this.walletRepos[0]) === null || _this$walletRepos$2 === void 0 ? void 0 : _this$walletRepos$2.wallets.forEach(function (wallet) {
        var _wallet$walletInfo$co3, _wallet$walletInfo$co4;
        (_wallet$walletInfo$co3 = wallet.walletInfo.connectEventNamesOnWindow) === null || _wallet$walletInfo$co3 === void 0 ? void 0 : _wallet$walletInfo$co3.forEach(function (eventName) {
          window.removeEventListener(eventName, _this._handleConnect);
        });
        (_wallet$walletInfo$co4 = wallet.walletInfo.connectEventNamesOnClient) === null || _wallet$walletInfo$co4 === void 0 ? void 0 : _wallet$walletInfo$co4.forEach( /*#__PURE__*/function () {
          var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(eventName) {
            var _ref5, _ref5$off;
            return _regenerator["default"].wrap(function _callee3$(_context3) {
              while (1) {
                switch (_context3.prev = _context3.next) {
                  case 0:
                    _context3.t1 = wallet.client;
                    if (_context3.t1) {
                      _context3.next = 5;
                      break;
                    }
                    _context3.next = 4;
                    return wallet.clientPromise;
                  case 4:
                    _context3.t1 = _context3.sent;
                  case 5:
                    _context3.t2 = _ref5 = _context3.t1;
                    _context3.t0 = _context3.t2 === null;
                    if (_context3.t0) {
                      _context3.next = 9;
                      break;
                    }
                    _context3.t0 = _ref5 === void 0;
                  case 9:
                    if (!_context3.t0) {
                      _context3.next = 13;
                      break;
                    }
                    void 0;
                    _context3.next = 14;
                    break;
                  case 13:
                    (_ref5$off = _ref5.off) === null || _ref5$off === void 0 ? void 0 : _ref5$off.call(_ref5, eventName, _this._handleConnect);
                  case 14:
                  case "end":
                    return _context3.stop();
                }
              }
            }, _callee3);
          }));
          return function (_x2) {
            return _ref4.apply(this, arguments);
          };
        }());
      });
    });
    _this.sessionOptions = _objectSpread(_objectSpread({}, _this.sessionOptions), sessionOptions);
    console.info("".concat(chains.length, " chains and ").concat(wallets.length, " wallets are provided!"));
    _this.chainRecords = chains.map(function (chain) {
      return (0, _utils.convertChain)(chain, assetLists, signerOptions, endpointOptions === null || endpointOptions === void 0 ? void 0 : endpointOptions[chain.chain_name]);
    });
    wallets.forEach(function (wallet) {
      wallet.setChains(_this.chainRecords);
    });
    if (_this.options.synchroMutexWallet) {
      wallets.forEach(function (_ref6) {
        var chainWallets = _ref6.chainWallets;
        chainWallets === null || chainWallets === void 0 ? void 0 : chainWallets.forEach(function (w) {
          w.updateCallbacks({
            afterDisconnect: function () {
              var _afterDisconnect = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
                return _regenerator["default"].wrap(function _callee5$(_context5) {
                  while (1) {
                    switch (_context5.prev = _context5.next) {
                      case 0:
                        chainWallets.forEach( /*#__PURE__*/function () {
                          var _ref7 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(w2) {
                            return _regenerator["default"].wrap(function _callee4$(_context4) {
                              while (1) {
                                switch (_context4.prev = _context4.next) {
                                  case 0:
                                    if (!(!w2.isWalletDisconnected && w2 !== w)) {
                                      _context4.next = 3;
                                      break;
                                    }
                                    _context4.next = 3;
                                    return w2.disconnect();
                                  case 3:
                                  case "end":
                                    return _context4.stop();
                                }
                              }
                            }, _callee4);
                          }));
                          return function (_x3) {
                            return _ref7.apply(this, arguments);
                          };
                        }());
                      case 1:
                      case "end":
                        return _context5.stop();
                    }
                  }
                }, _callee5);
              }));
              function afterDisconnect() {
                return _afterDisconnect.apply(this, arguments);
              }
              return afterDisconnect;
            }(),
            afterConnect: function () {
              var _afterConnect = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
                return _regenerator["default"].wrap(function _callee6$(_context6) {
                  while (1) {
                    switch (_context6.prev = _context6.next) {
                      case 0:
                        _this.synchronizeMutexWalletConnection();
                      case 1:
                      case "end":
                        return _context6.stop();
                    }
                  }
                }, _callee6);
              }));
              function afterConnect() {
                return _afterConnect.apply(this, arguments);
              }
              return afterConnect;
            }()
          });
        });
      });
    }
    _this.chainRecords.map(function (chainRecord) {
      _this.walletRepos.push(new _repository.WalletRepo(chainRecord, wallets.map(function (_ref8) {
        var getChainWallet = _ref8.getChainWallet;
        return getChainWallet(chainRecord.name);
      }), _this.sessionOptions));
    });
    return _this;
  }
  (0, _createClass2["default"])(WalletManagerV2, [{
    key: "walletReposInUse",
    get: function get() {
      return this.walletRepos.filter(function (repo) {
        return repo.isInUse === true;
      });
    }
  }, {
    key: "synchronizeMutexWalletConnection",
    value: function () {
      var _synchronizeMutexWalletConnection = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7() {
        var ls, walletName, _iterator, _step, repo, _repo$current, _repo$current2;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                if (!(typeof window === 'undefined')) {
                  _context7.next = 2;
                  break;
                }
                return _context7.abrupt("return");
              case 2:
                ls = window.localStorage;
                if (!(ls.getItem('synchronize-mutex-wallet') === 'done')) {
                  _context7.next = 5;
                  break;
                }
                return _context7.abrupt("return");
              case 5:
                ls.setItem('synchronize-mutex-wallet', 'done');
                walletName = ls.getItem('chain-provider');
                _iterator = _createForOfIteratorHelper(this.walletReposInUse);
                _context7.prev = 8;
                _iterator.s();
              case 10:
                if ((_step = _iterator.n()).done) {
                  _context7.next = 20;
                  break;
                }
                repo = _step.value;
                if (!walletName) {
                  _context7.next = 18;
                  break;
                }
                if (!(((_repo$current = repo.current) === null || _repo$current === void 0 ? void 0 : _repo$current.walletName) !== walletName)) {
                  _context7.next = 16;
                  break;
                }
                _context7.next = 16;
                return (_repo$current2 = repo.current) === null || _repo$current2 === void 0 ? void 0 : _repo$current2.disconnect();
              case 16:
                _context7.next = 18;
                return repo.connect(walletName);
              case 18:
                _context7.next = 10;
                break;
              case 20:
                _context7.next = 25;
                break;
              case 22:
                _context7.prev = 22;
                _context7.t0 = _context7["catch"](8);
                _iterator.e(_context7.t0);
              case 25:
                _context7.prev = 25;
                _iterator.f();
                return _context7.finish(25);
              case 28:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this, [[8, 22, 25, 28]]);
      }));
      function synchronizeMutexWalletConnection() {
        return _synchronizeMutexWalletConnection.apply(this, arguments);
      }
      return synchronizeMutexWalletConnection;
    }()
  }]);
  return WalletManagerV2;
}(_bases.StateBase);
exports.WalletManagerV2 = WalletManagerV2;
