// @cosmos-kit/keplr/main/wallet-connect/client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeplrClient = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _client = _interopRequireDefault(require("@walletconnect/client"));
var _utils = require("@walletconnect/utils");
var _deepmerge = _interopRequireDefault(require("deepmerge"));
/* eslint-disable no-console */
var KeplrClient = /*#__PURE__*/function () {
  function KeplrClient() {
    (0, _classCallCheck2["default"])(this, KeplrClient);
    (0, _defineProperty2["default"])(this, "defaultOptions", {});
    (0, _defineProperty2["default"])(this, "connector", void 0);
    this.connector = new _client["default"]({
      bridge: 'https://bridge.walletconnect.org',
      signingMethods: ['keplr_enable_wallet_connect_v1', 'keplr_get_key_wallet_connect_v1', 'keplr_sign_amino_wallet_connect_v1']
    });
  }
  (0, _createClass2["default"])(KeplrClient, [{
    key: "qrUrl",
    get: function get() {
      return this.connector.uri;
    }
  }, {
    key: "getAppUrl",
    value: function getAppUrl(os) {
      switch (os) {
        case 'android':
          (0, _utils.saveMobileLinkInfo)({
            name: 'Keplr',
            href: 'intent://wcV1#Intent;package=com.chainapsis.keplr;scheme=keplrwallet;end;'
          });
          return "intent://wcV1?".concat(this.qrUrl, "#Intent;package=com.chainapsis.keplr;scheme=keplrwallet;end;");
        case 'ios':
          (0, _utils.saveMobileLinkInfo)({
            name: 'Keplr',
            href: 'keplrwallet://wcV1'
          });
          return "keplrwallet://wcV1?".concat(this.qrUrl);
        default:
          return void 0;
      }
    }
  }, {
    key: "enable",
    value: function () {
      var _enable = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(chainIds) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (typeof chainIds === 'string') {
                  chainIds = [chainIds];
                }
                _context.next = 3;
                return this.connector.sendCustomRequest({
                  id: (0, _utils.payloadId)(),
                  jsonrpc: '2.0',
                  method: 'keplr_enable_wallet_connect_v1',
                  params: chainIds
                });
              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
      function enable(_x) {
        return _enable.apply(this, arguments);
      }
      return enable;
    }()
  }, {
    key: "getAccount",
    value: function () {
      var _getAccount = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(chainId) {
        var response;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.connector.sendCustomRequest({
                  id: (0, _utils.payloadId)(),
                  jsonrpc: '2.0',
                  method: 'keplr_get_key_wallet_connect_v1',
                  params: [chainId]
                });
              case 2:
                response = _context2.sent[0];
                return _context2.abrupt("return", {
                  name: response.name,
                  address: response.bech32Address,
                  algo: response.algo,
                  pubkey: Buffer.from(response.pubKey, 'hex')
                });
              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));
      function getAccount(_x2) {
        return _getAccount.apply(this, arguments);
      }
      return getAccount;
    }()
  }, {
    key: "getOfflineSigner",
    value: function getOfflineSigner(chainId) {
      var _this = this;
      return {
        getAccounts: function () {
          var _getAccounts = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
            return _regenerator["default"].wrap(function _callee3$(_context3) {
              while (1) {
                switch (_context3.prev = _context3.next) {
                  case 0:
                    _context3.next = 2;
                    return _this.getAccount(chainId);
                  case 2:
                    _context3.t0 = _context3.sent;
                    return _context3.abrupt("return", [_context3.t0]);
                  case 4:
                  case "end":
                    return _context3.stop();
                }
              }
            }, _callee3);
          }));
          function getAccounts() {
            return _getAccounts.apply(this, arguments);
          }
          return getAccounts;
        }(),
        signAmino: function signAmino(signerAddress, signDoc) {
          return _this.signAmino(chainId, signerAddress, signDoc);
        }
      };
    }
  }, {
    key: "signAmino",
    value: function () {
      var _signAmino = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(chainId, signer, signDoc) {
        var _this$defaultOptions$;
        var signOptions,
          _args4 = arguments;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                signOptions = _args4.length > 3 && _args4[3] !== undefined ? _args4[3] : {};
                _context4.next = 3;
                return this.connector.sendCustomRequest({
                  id: (0, _utils.payloadId)(),
                  jsonrpc: '2.0',
                  method: 'keplr_sign_amino_wallet_connect_v1',
                  params: [chainId, signer, signDoc, (0, _deepmerge["default"])((_this$defaultOptions$ = this.defaultOptions.sign) !== null && _this$defaultOptions$ !== void 0 ? _this$defaultOptions$ : {}, signOptions)]
                });
              case 3:
                return _context4.abrupt("return", _context4.sent[0]);
              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));
      function signAmino(_x3, _x4, _x5) {
        return _signAmino.apply(this, arguments);
      }
      return signAmino;
    }()
  }]);
  return KeplrClient;
}();
exports.KeplrClient = KeplrClient;
