// @cosmos-kit/keplr/main/extension/client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeplrClient = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _keplr = require("@chain-registry/keplr");
var KeplrClient = /*#__PURE__*/function () {
  function KeplrClient(client) {
    (0, _classCallCheck2["default"])(this, KeplrClient);
    (0, _defineProperty2["default"])(this, "client", void 0);
    this.client = client;
  }
  (0, _createClass2["default"])(KeplrClient, [{
    key: "enable",
    value: function () {
      var _enable = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(chainIds) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.client.enable(chainIds);
              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
      function enable(_x) {
        return _enable.apply(this, arguments);
      }
      return enable;
    }()
  }, {
    key: "getAccount",
    value: function () {
      var _getAccount = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(chainId) {
        var key;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.client.getKey(chainId);
              case 2:
                key = _context2.sent;
                return _context2.abrupt("return", {
                  name: key.name,
                  address: key.bech32Address,
                  algo: key.algo,
                  pubkey: key.pubKey
                });
              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));
      function getAccount(_x2) {
        return _getAccount.apply(this, arguments);
      }
      return getAccount;
    }()
  }, {
    key: "getOfflineSigner",
    value: function getOfflineSigner(chainId) {
      return this.client.getOfflineSignerAuto(chainId);
    }
  }, {
    key: "getOfflineSignerAmino",
    value: function getOfflineSignerAmino(chainId) {
      return this.client.getOfflineSignerOnlyAmino(chainId);
    }
  }, {
    key: "getOfflineSignerDirect",
    value: function getOfflineSignerDirect(chainId) {
      return this.client.getOfflineSigner(chainId);
    }
  }, {
    key: "addChain",
    value: function () {
      var _addChain = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(chainInfo) {
        var _chainInfo$preferredE, _chainInfo$preferredE2, _chainInfo$preferredE5, _chainInfo$preferredE6;
        var suggestChain, _chainInfo$preferredE3, _chainInfo$preferredE4, _chainInfo$preferredE7, _chainInfo$preferredE8;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                suggestChain = (0, _keplr.chainRegistryChainToKeplr)(chainInfo.chain, chainInfo.assetList ? [chainInfo.assetList] : []);
                if ((_chainInfo$preferredE = chainInfo.preferredEndpoints) !== null && _chainInfo$preferredE !== void 0 && (_chainInfo$preferredE2 = _chainInfo$preferredE.rest) !== null && _chainInfo$preferredE2 !== void 0 && _chainInfo$preferredE2[0]) {
                  suggestChain.rest = (_chainInfo$preferredE3 = chainInfo.preferredEndpoints) === null || _chainInfo$preferredE3 === void 0 ? void 0 : (_chainInfo$preferredE4 = _chainInfo$preferredE3.rest) === null || _chainInfo$preferredE4 === void 0 ? void 0 : _chainInfo$preferredE4[0];
                }
                if ((_chainInfo$preferredE5 = chainInfo.preferredEndpoints) !== null && _chainInfo$preferredE5 !== void 0 && (_chainInfo$preferredE6 = _chainInfo$preferredE5.rpc) !== null && _chainInfo$preferredE6 !== void 0 && _chainInfo$preferredE6[0]) {
                  suggestChain.rpc = (_chainInfo$preferredE7 = chainInfo.preferredEndpoints) === null || _chainInfo$preferredE7 === void 0 ? void 0 : (_chainInfo$preferredE8 = _chainInfo$preferredE7.rpc) === null || _chainInfo$preferredE8 === void 0 ? void 0 : _chainInfo$preferredE8[0];
                }
                _context3.next = 5;
                return this.client.experimentalSuggestChain(suggestChain);
              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));
      function addChain(_x3) {
        return _addChain.apply(this, arguments);
      }
      return addChain;
    }()
  }, {
    key: "signAmino",
    value: function () {
      var _signAmino = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(chainId, signer, signDoc, signOptions) {
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.client.signAmino(chainId, signer, signDoc, signOptions);
              case 2:
                return _context4.abrupt("return", _context4.sent);
              case 3:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));
      function signAmino(_x4, _x5, _x6, _x7) {
        return _signAmino.apply(this, arguments);
      }
      return signAmino;
    }()
  }, {
    key: "signDirect",
    value: function () {
      var _signDirect = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(chainId, signer, signDoc, signOptions) {
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this.client.signDirect(chainId, signer, signDoc, signOptions);
              case 2:
                return _context5.abrupt("return", _context5.sent);
              case 3:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));
      function signDirect(_x8, _x9, _x10, _x11) {
        return _signDirect.apply(this, arguments);
      }
      return signDirect;
    }()
  }, {
    key: "sendTx",
    value: function () {
      var _sendTx = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(chainId, tx, mode) {
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this.client.sendTx(chainId, tx, mode);
              case 2:
                return _context6.abrupt("return", _context6.sent);
              case 3:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));
      function sendTx(_x12, _x13, _x14) {
        return _sendTx.apply(this, arguments);
      }
      return sendTx;
    }()
  }]);
  return KeplrClient;
}();
exports.KeplrClient = KeplrClient;
