// @cosmos-kit/keplr/main/extension/registry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keplrExtensionInfo = void 0;
var _go = require("react-icons/go");
var _gr = require("react-icons/gr");
var _ri = require("react-icons/ri");
var keplrExtensionInfo = {
  name: 'keplr-extension',
  logo: 'https://pbs.twimg.com/profile_images/1498228570862219266/uctq7aeh_400x400.png',
  prettyName: 'Keplr',
  mode: 'extension',
  mobileDisabled: true,
  rejectMessage: {
    source: 'Request rejected'
  },
  connectEventNamesOnWindow: ['keplr_keystorechange'],
  downloads: [{
    device: 'desktop',
    browser: 'chrome',
    icon: _ri.RiChromeFill,
    link: 'https://chrome.google.com/webstore/detail/keplr/dmkamcknogkgcdfhhbddcghachkejeap?hl=en'
  }, {
    device: 'desktop',
    browser: 'firefox',
    icon: _gr.GrFirefox,
    link: 'https://addons.mozilla.org/en-US/firefox/addon/keplr/'
  }, {
    icon: _go.GoDesktopDownload,
    link: 'https://www.keplr.app/download'
  }]
};
exports.keplrExtensionInfo = keplrExtensionInfo;
