// @cosmos-kit/react/main/modal/components/motion-component.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalContentVariants = exports.LoadingVariants = exports.AnimateGridItem = exports.AnimateBox = void 0;
var _react = require("@chakra-ui/react");
var _framerMotion = require("framer-motion");
var ModalContentVariants = {
  hidden: {
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
    height: '10%',
    opacity: 0,
    scale: 0.8,
    transition: {
      duration: 0.4,
      type: 'easeOut'
    }
  },
  enter: {
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
    height: '100%',
    opacity: 1,
    scale: 1,
    transition: {
      duration: 0.5,
      type: 'easeOut'
    }
  }
};
exports.ModalContentVariants = ModalContentVariants;
var LoadingVariants = {
  hidden: {
    rotate: 0,
    transition: {
      duration: 0.4,
      type: 'easeOut'
    }
  },
  animate: {
    opacity: [1, 0.7, 0.5, 0.3, 0.5, 0.7, 1],
    rotate: 360,
    transition: {
      duration: 2,
      type: 'easeInOut',
      repeat: Infinity,
      repeatType: 'loop'
    }
  }
};
exports.LoadingVariants = LoadingVariants;
var AnimateBox = (0, _framerMotion.motion)(_react.Box);
exports.AnimateBox = AnimateBox;
var AnimateGridItem = (0, _framerMotion.motion)(_react.GridItem);
exports.AnimateGridItem = AnimateGridItem;
