// @cosmos-kit/react/main/modal/components/SimpleConnectModal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleConnectModal = void 0;
var _react = require("@chakra-ui/react");
var _react2 = _interopRequireDefault(require("react"));
var SimpleConnectModal = function SimpleConnectModal(_ref) {
  var initialRef = _ref.initialRef,
    modalHead = _ref.modalHead,
    modalContent = _ref.modalContent,
    modalIsOpen = _ref.modalIsOpen,
    modalOnClose = _ref.modalOnClose;
  return /*#__PURE__*/_react2["default"].createElement(_react.Modal, {
    initialFocusRef: initialRef,
    isOpen: modalIsOpen,
    isCentered: true,
    onClose: modalOnClose
  }, /*#__PURE__*/_react2["default"].createElement(_react.ModalOverlay, null), /*#__PURE__*/_react2["default"].createElement(_react.ModalContent, {
    position: "relative",
    alignSelf: "center",
    borderRadius: "xl",
    w: "fit-content",
    pb: 4,
    mx: 4,
    _focus: {
      outline: 'none'
    },
    overflow: "hidden"
  }, /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    flex: 1,
    spacing: 1,
    h: "full"
  }, modalHead, modalContent)));
};
exports.SimpleConnectModal = SimpleConnectModal;
