// @cosmos-kit/react/main/modal/components/SimpleConnectModalContentV1.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleQRCodeV1 = exports.SimpleInstallWalletButtonV1 = exports.SimpleDisplayWalletListV1 = exports.SimpleDisplayModalContentV1 = void 0;
var _react = require("@chakra-ui/react");
var _qrcode = require("qrcode.react");
var _react2 = _interopRequireWildcard(require("react"));
var _fi = require("react-icons/fi");
var _defaultComponent = require("./default-component");
var _motionComponent = require("./motion-component");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* eslint-disable @typescript-eslint/no-non-null-assertion */

var SimpleInstallWalletButtonV1 = function SimpleInstallWalletButtonV1(_ref) {
  var icon = _ref.icon,
    text = _ref.text,
    disabled = _ref.disabled,
    onClick = _ref.onClick;
  var _useColorMode = (0, _react.useColorMode)(),
    colorMode = _useColorMode.colorMode;
  return /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    w: "full",
    px: 6,
    mt: 1
  }, /*#__PURE__*/_react2["default"].createElement(_react.Button, {
    variant: "unstyled",
    w: "full",
    h: "auto",
    fontWeight: "medium",
    fontSize: "lg",
    color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'rgba(37, 57, 201, 0.72)', 'rgba(196, 203, 255, 0.9)'),
    border: "1px solid",
    borderColor: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, '#ffffff', 'rgba(0, 0, 0, 0.25)'),
    bg: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'rgba(37, 57, 201, 0.1)', 'rgba(40, 62, 219, 0.15)'),
    boxShadow: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, '0 0 1px 2px rgba(37, 57, 201, 0.5)', '0 0 1px 2px rgba(196, 203, 255, 0.5)'),
    _hover: {
      opacity: 0.8
    },
    _active: {
      opacity: 0.9
    },
    _focus: {
      outline: 'none'
    },
    _disabled: {
      opacity: 0.5,
      cursor: 'not-allowed',
      _hover: {
        opacity: 0.5
      },
      _active: {
        opacity: 0.5
      }
    },
    isDisabled: disabled,
    onClick: onClick
  }, /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    w: "full",
    isInline: true,
    justifyContent: "center",
    alignItems: "center",
    p: 3
  }, icon && /*#__PURE__*/_react2["default"].createElement(_react.Icon, {
    as: icon
  }), /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    whiteSpace: "break-spaces"
  }, text ? text : "Install Wallet"))));
};
exports.SimpleInstallWalletButtonV1 = SimpleInstallWalletButtonV1;
var SimpleDisplayModalContentV1 = function SimpleDisplayModalContentV1(_ref2) {
  var _Style;
  var status = _ref2.status,
    logo = _ref2.logo,
    contentHeader = _ref2.contentHeader,
    contentDesc = _ref2.contentDesc,
    username = _ref2.username,
    walletIcon = _ref2.walletIcon,
    addressButton = _ref2.addressButton,
    bottomButton = _ref2.bottomButton;
  var _useColorMode2 = (0, _react.useColorMode)(),
    colorMode = _useColorMode2.colorMode;
  var Style = {
    warning: {
      color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'orange.300', 'orange.400')
    },
    error: {
      color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'red.400', 'red.500')
    },
    loading: {
      color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'inherit', 'inherit')
    }
  };
  return /*#__PURE__*/_react2["default"].createElement(_motionComponent.AnimateBox, {
    initial: "hidden",
    animate: "enter",
    exit: "exit",
    variants: _motionComponent.ModalContentVariants
  }, /*#__PURE__*/_react2["default"].createElement(_react.Flex, {
    w: 96,
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    textAlign: "center",
    p: 4,
    pt: 10
  }, logo && /*#__PURE__*/_react2["default"].createElement(_react.Center, {
    position: "relative",
    mx: "auto",
    w: 24,
    h: 24,
    minW: 24,
    minH: 24,
    maxW: 24,
    maxH: 24,
    mb: typeof logo === 'string' ? 6 : 0
  }, status === 'loading' && /*#__PURE__*/_react2["default"].createElement(_motionComponent.AnimateBox, {
    position: "absolute",
    top: -2,
    right: -2,
    bottom: -2,
    left: -2,
    border: "2px solid",
    borderTopColor: "transparent",
    borderBottomColor: "transparent",
    borderLeftColor: "purple.300",
    borderRightColor: "purple.300",
    borderRadius: "full",
    initial: "hidden",
    animate: "animate",
    variants: _motionComponent.LoadingVariants
  }), (status === 'warning' || status === 'error') && /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    position: "absolute",
    top: -2,
    right: -2,
    bottom: -2,
    left: -2,
    border: "2px solid",
    borderColor: Style[status].color,
    borderRadius: "full"
  }), /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    borderRadius: "full",
    overflow: "hidden"
  }, typeof logo === 'string' ? /*#__PURE__*/_react2["default"].createElement(_react.Image, {
    src: logo
  }) : /*#__PURE__*/_react2["default"].createElement(_react.Icon, {
    as: logo,
    w: "full",
    h: "full"
  }))), contentHeader && /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    fontSize: "lg",
    fontWeight: "semibold",
    color: (_Style = Style[status]) === null || _Style === void 0 ? void 0 : _Style.color,
    mb: 0.5
  }, contentHeader), contentDesc && /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    lineHeight: 1.3,
    opacity: 0.7,
    whiteSpace: "pre-line"
  }, contentDesc), username && /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    isInline: true,
    justifyContent: "center",
    alignItems: "center"
  }, /*#__PURE__*/_react2["default"].createElement(_react.Center, {
    w: 5,
    h: 5,
    minW: 5,
    minH: 5,
    maxW: 5,
    maxH: 5
  }, /*#__PURE__*/_react2["default"].createElement(_react.Image, {
    src: walletIcon
  })), /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    fontSize: "xl",
    fontWeight: "semibold"
  }, username)), addressButton && /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    w: "full",
    pt: 4,
    px: 8
  }, addressButton), bottomButton && /*#__PURE__*/_react2["default"].createElement(_react.Box, {
    w: "full",
    pt: addressButton ? 4 : 4
  }, bottomButton)));
};
exports.SimpleDisplayModalContentV1 = SimpleDisplayModalContentV1;
var SimpleQRCodeV1 = function SimpleQRCodeV1(_ref3) {
  var link = _ref3.link,
    description = _ref3.description;
  var elementRef = (0, _react2.useRef)(null);
  var dimensions = (0, _react.useDimensions)(elementRef);
  var _useColorMode3 = (0, _react.useColorMode)(),
    colorMode = _useColorMode3.colorMode;
  return /*#__PURE__*/_react2["default"].createElement(_motionComponent.AnimateBox, {
    ref: elementRef,
    initial: "hidden",
    animate: "enter",
    exit: "exit",
    variants: _motionComponent.ModalContentVariants
  }, /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    w: 96,
    justifyContent: "center",
    alignItems: "center",
    spacing: 4,
    p: 8,
    pt: 6
  }, description && /*#__PURE__*/_react2["default"].createElement(_react.Text, {
    fontWeight: "medium",
    textAlign: "center",
    opacity: 0.75
  }, description), /*#__PURE__*/_react2["default"].createElement(_react.Center, {
    w: "full",
    border: "1px solid",
    borderColor: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'blackAlpha.100', 'whiteAlpha.600'),
    borderRadius: "lg",
    boxShadow: "base",
    p: 5
  }, /*#__PURE__*/_react2["default"].createElement(_qrcode.QRCodeSVG, {
    value: link,
    size: dimensions ? dimensions.contentBox.width - 24 : void 0,
    bgColor: '#ffffff',
    fgColor: '#000000',
    level: 'L',
    includeMargin: false
  }))));
};
exports.SimpleQRCodeV1 = SimpleQRCodeV1;
var SimpleDisplayWalletListV1 = function SimpleDisplayWalletListV1(_ref4) {
  var initialFocus = _ref4.initialFocus,
    walletsData = _ref4.walletsData;
  var _useColorMode4 = (0, _react.useColorMode)(),
    colorMode = _useColorMode4.colorMode;
  return /*#__PURE__*/_react2["default"].createElement(_motionComponent.AnimateBox, {
    initial: "hidden",
    animate: "enter",
    variants: _motionComponent.ModalContentVariants
  }, /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    flex: 1,
    spacing: 3,
    maxH: 96,
    minH: 60,
    w: 96,
    overflowY: "scroll",
    pl: 4,
    pr: 1,
    py: 0.5,
    css: {
      // For Firefox
      scrollbarWidth: 'auto',
      scrollbarColor: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'rgba(0,0,0,0.3) rgba(0,0,0,0.2)', 'rgba(255,255,255,0.2) rgba(255,255,255,0.1)'),
      // For Chrome and other browsers except Firefox
      '&::-webkit-scrollbar': {
        width: '10px',
        background: 'transparent',
        // background: "gray",
        borderRadius: '3px',
        mr: 1
      },
      '&::-webkit-scrollbar-thumb': {
        background: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'rgba(0,0,0,0.1)', 'rgba(255,255,255,0.1)'),
        borderRadius: '6px',
        border: '3px solid transparent',
        backgroundClip: 'content-box'
      }
    }
  }, walletsData.map(function (_ref5, i) {
    var name = _ref5.name,
      prettyName = _ref5.prettyName,
      logo = _ref5.logo,
      onClick = _ref5.onClick;
    return /*#__PURE__*/_react2["default"].createElement(_react.Button, {
      ref: i === 0 ? initialFocus : null,
      id: name,
      key: name,
      variant: "unstyled",
      display: "flex",
      h: "fit-content",
      p: 2.5,
      justifyContent: "start",
      borderRadius: "none",
      whiteSpace: "break-spaces",
      color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'blackAlpha.800', 'whiteAlpha.800'),
      boxShadow: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, '0 20px 1px -19px #fff', '0 20px 1px -19px #2d3748'),
      transition: "all .4s ease-in-out",
      _hover: {
        color: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, 'primary.300', 'primary.100'),
        borderRadius: 'md',
        boxShadow: (0, _defaultComponent.handleChangeColorModeValue)(colorMode, '0 0 2px 0 rgba(98, 17, 240, 0.5)', '0 0 2px 0 rgba(182, 153, 232, 0.9)')
      },
      _focus: {
        borderRadius: 'md',
        outline: 'none'
      },
      onClick: onClick
    }, /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
      w: "full",
      isInline: true,
      justifyContent: "start",
      alignItems: "center",
      spacing: 2.5
    }, /*#__PURE__*/_react2["default"].createElement(_react.Box, {
      borderRadius: "lg",
      overflow: "hidden",
      w: 9,
      h: 9,
      minW: 9,
      minH: 9,
      maxW: 9,
      maxH: 9
    }, /*#__PURE__*/_react2["default"].createElement(_react.Image, {
      src: typeof logo === 'string' ? logo : void 0
    })), /*#__PURE__*/_react2["default"].createElement(_react.Box, {
      textAlign: "start",
      flex: 1
    }, /*#__PURE__*/_react2["default"].createElement(_react.Text, {
      fontSize: "xl",
      fontWeight: "semibold",
      lineHeight: 1.1
    }, prettyName)), /*#__PURE__*/_react2["default"].createElement(_react.Box, null, /*#__PURE__*/_react2["default"].createElement(_react.Icon, {
      as: _fi.FiChevronRight
    }))));
  })));
};
exports.SimpleDisplayWalletListV1 = SimpleDisplayWalletListV1;
