// @cosmos-kit/react/main/modal/components/SimpleConnectModalV1.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleConnectModalV1 = void 0;
var _react = require("@chakra-ui/react");
var _react2 = _interopRequireDefault(require("react"));
var _motionComponent = require("./motion-component");
var SimpleConnectModalV1 = function SimpleConnectModalV1(_ref) {
  var initialRef = _ref.initialRef,
    modalHead = _ref.modalHead,
    modalContent = _ref.modalContent,
    modalIsOpen = _ref.modalIsOpen,
    modalOnClose = _ref.modalOnClose;
  return /*#__PURE__*/_react2["default"].createElement(_react.Modal, {
    initialFocusRef: initialRef,
    isOpen: modalIsOpen,
    isCentered: true,
    onClose: modalOnClose
  }, /*#__PURE__*/_react2["default"].createElement(_react.ModalOverlay, null), /*#__PURE__*/_react2["default"].createElement(_react.ModalContent, {
    alignSelf: "center",
    borderRadius: "xl",
    w: "full",
    maxW: "fit-content",
    p: 1,
    pb: 8,
    mx: 4,
    _focus: {
      outline: 'none'
    }
  }, /*#__PURE__*/_react2["default"].createElement(_motionComponent.AnimateBox, {
    initial: "hidden",
    animate: "enter",
    variants: _motionComponent.ModalContentVariants
  }, /*#__PURE__*/_react2["default"].createElement(_react.Stack, {
    flex: 1,
    spacing: 0,
    h: "full"
  }, modalHead, modalContent))));
};
exports.SimpleConnectModalV1 = SimpleConnectModalV1;
