// @cosmos-kit/core/main/utils/convert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.convertChain = convertChain;
function convertChain(chain, assetLists, signerOptions, preferredEndpoints) {
  var _signerOptions$starga, _signerOptions$signin, _signerOptions$signin2;
  var assetList = assetLists.find(function (list) {
    return list.chain_name === chain.chain_name;
  });
  return {
    name: chain.chain_name,
    chain: chain,
    assetList: assetList,
    clientOptions: {
      stargate: signerOptions === null || signerOptions === void 0 ? void 0 : (_signerOptions$starga = signerOptions.stargate) === null || _signerOptions$starga === void 0 ? void 0 : _signerOptions$starga.call(signerOptions, chain),
      signingStargate: signerOptions === null || signerOptions === void 0 ? void 0 : (_signerOptions$signin = signerOptions.signingStargate) === null || _signerOptions$signin === void 0 ? void 0 : _signerOptions$signin.call(signerOptions, chain),
      signingCosmwasm: signerOptions === null || signerOptions === void 0 ? void 0 : (_signerOptions$signin2 = signerOptions.signingCosmwasm) === null || _signerOptions$signin2 === void 0 ? void 0 : _signerOptions$signin2.call(signerOptions, chain)
    },
    preferredEndpoints: preferredEndpoints
  };
}
