// @cosmos-kit/cosmostation/main/extension/client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CosmostationClient = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _cosmostation = require("@chain-registry/cosmostation");
var CosmostationClient = /*#__PURE__*/function () {
  function CosmostationClient(client) {
    (0, _classCallCheck2["default"])(this, CosmostationClient);
    (0, _defineProperty2["default"])(this, "client", void 0);
    (0, _defineProperty2["default"])(this, "eventMap", new Map());
    this.client = client;
  }
  (0, _createClass2["default"])(CosmostationClient, [{
    key: "cosmos",
    get: function get() {
      return this.client.cosmos;
    }
  }, {
    key: "ikeplr",
    get: function get() {
      return this.client.providers.keplr;
    }
  }, {
    key: "getAccount",
    value: function () {
      var _getAccount = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(chainId) {
        var key;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.cosmos.request({
                  method: 'cos_requestAccount',
                  params: {
                    chainName: chainId
                  }
                });
              case 2:
                key = _context.sent;
                return _context.abrupt("return", {
                  name: key.name,
                  address: key.address,
                  pubkey: key.publicKey
                });
              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
      function getAccount(_x) {
        return _getAccount.apply(this, arguments);
      }
      return getAccount;
    }()
  }, {
    key: "disconnect",
    value: function () {
      var _disconnect = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.cosmos.request({
                  method: 'cos_disconnect'
                });
              case 2:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));
      function disconnect() {
        return _disconnect.apply(this, arguments);
      }
      return disconnect;
    }()
  }, {
    key: "on",
    value: function on(type, listener) {
      var event = this.cosmos.on(type, listener);
      var typeEventMap = this.eventMap.get(type) || new Map();
      typeEventMap.set(listener, event);
      this.eventMap.set(type, typeEventMap);
    }
  }, {
    key: "off",
    value: function off(type, listener) {
      var _this$eventMap$get;
      var event = (_this$eventMap$get = this.eventMap.get(type)) === null || _this$eventMap$get === void 0 ? void 0 : _this$eventMap$get.get(listener);
      if (event) {
        this.cosmos.off(event);
      }
    }
  }, {
    key: "getOfflineSigner",
    value: function getOfflineSigner(chainId) {
      return this.ikeplr.getOfflineSignerAuto(chainId);
    }
  }, {
    key: "getOfflineSignerAmino",
    value: function getOfflineSignerAmino(chainId) {
      return this.ikeplr.getOfflineSignerOnlyAmino(chainId);
    }
  }, {
    key: "getOfflineSignerDirect",
    value: function getOfflineSignerDirect(chainId) {
      return this.ikeplr.getOfflineSigner(chainId);
    }
  }, {
    key: "addChain",
    value: function () {
      var _addChain = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(chainInfo) {
        var _chainInfo$preferredE, _chainInfo$preferredE2;
        var suggestChain, _chainInfo$preferredE3, _chainInfo$preferredE4, result;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                suggestChain = (0, _cosmostation.chainRegistryChainToCosmostation)(chainInfo.chain, chainInfo.assetList ? [chainInfo.assetList] : []);
                if ((_chainInfo$preferredE = chainInfo.preferredEndpoints) !== null && _chainInfo$preferredE !== void 0 && (_chainInfo$preferredE2 = _chainInfo$preferredE.rest) !== null && _chainInfo$preferredE2 !== void 0 && _chainInfo$preferredE2[0]) {
                  suggestChain.restURL = (_chainInfo$preferredE3 = chainInfo.preferredEndpoints) === null || _chainInfo$preferredE3 === void 0 ? void 0 : (_chainInfo$preferredE4 = _chainInfo$preferredE3.rest) === null || _chainInfo$preferredE4 === void 0 ? void 0 : _chainInfo$preferredE4[0];
                }
                _context3.next = 4;
                return this.cosmos.request({
                  method: 'cos_addChain',
                  params: suggestChain
                });
              case 4:
                result = _context3.sent;
                if (result) {
                  _context3.next = 7;
                  break;
                }
                throw new Error("Failed to add chain ".concat(chainInfo.name, "."));
              case 7:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));
      function addChain(_x2) {
        return _addChain.apply(this, arguments);
      }
      return addChain;
    }()
  }, {
    key: "signAmino",
    value: function () {
      var _signAmino = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(chainId, signer, signDoc, signOptions) {
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.prev = 0;
                _context4.next = 3;
                return this.ikeplr.signAmino(chainId, signer, signDoc, signOptions);
              case 3:
                return _context4.abrupt("return", _context4.sent);
              case 6:
                _context4.prev = 6;
                _context4.t0 = _context4["catch"](0);
                _context4.next = 10;
                return this.cosmos.request({
                  method: 'cos_signAmino',
                  params: {
                    chainName: chainId,
                    doc: signDoc,
                    isEditMemo: signOptions === null || signOptions === void 0 ? void 0 : signOptions.preferNoSetMemo,
                    isEditFee: signOptions === null || signOptions === void 0 ? void 0 : signOptions.preferNoSetFee
                  }
                });
              case 10:
                return _context4.abrupt("return", _context4.sent);
              case 11:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this, [[0, 6]]);
      }));
      function signAmino(_x3, _x4, _x5, _x6) {
        return _signAmino.apply(this, arguments);
      }
      return signAmino;
    }()
  }, {
    key: "signDirect",
    value: function () {
      var _signDirect = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(chainId, signer, signDoc, signOptions) {
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.prev = 0;
                _context5.next = 3;
                return this.ikeplr.signDirect(chainId, signer, signDoc, signOptions);
              case 3:
                return _context5.abrupt("return", _context5.sent);
              case 6:
                _context5.prev = 6;
                _context5.t0 = _context5["catch"](0);
                _context5.next = 10;
                return this.cosmos.request({
                  method: 'cos_signDirect',
                  params: {
                    chainName: chainId,
                    doc: signDoc,
                    isEditMemo: signOptions === null || signOptions === void 0 ? void 0 : signOptions.preferNoSetMemo,
                    isEditFee: signOptions === null || signOptions === void 0 ? void 0 : signOptions.preferNoSetFee
                  }
                });
              case 10:
                return _context5.abrupt("return", _context5.sent);
              case 11:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this, [[0, 6]]);
      }));
      function signDirect(_x7, _x8, _x9, _x10) {
        return _signDirect.apply(this, arguments);
      }
      return signDirect;
    }()
  }, {
    key: "sendTx",
    value: function () {
      var _sendTx = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(chainId, tx, mode) {
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this.ikeplr.sendTx(chainId, tx, mode);
              case 2:
                return _context6.abrupt("return", _context6.sent);
              case 3:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));
      function sendTx(_x11, _x12, _x13) {
        return _sendTx.apply(this, arguments);
      }
      return sendTx;
    }()
  }]);
  return CosmostationClient;
}();
exports.CosmostationClient = CosmostationClient;
