// @cosmos-kit/core/main/bases/wallet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WalletBase = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _types = require("../types");
var _utils = require("../utils");
var _state = require("./state");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var WalletBase = /*#__PURE__*/function (_StateBase) {
  (0, _inherits2["default"])(WalletBase, _StateBase);
  var _super = _createSuper(WalletBase);
  function WalletBase(walletInfo) {
    var _this;
    (0, _classCallCheck2["default"])(this, WalletBase);
    _this = _super.call(this);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "clientPromise", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "client", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_walletInfo", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_appUrl", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_qrUrl", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "callbacks", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "disconnect", /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(callbacks) {
        var _ref2, _ref2$beforeDisconnec, _this$client, _this$client$disconne, _ref3, _ref3$afterDisconnect;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return (_ref2 = callbacks || _this.callbacks) === null || _ref2 === void 0 ? void 0 : (_ref2$beforeDisconnec = _ref2.beforeDisconnect) === null || _ref2$beforeDisconnec === void 0 ? void 0 : _ref2$beforeDisconnec.call(_ref2);
              case 2:
                _this.reset();
                window.localStorage.removeItem('chain-provider');
                _context.next = 6;
                return (_this$client = _this.client) === null || _this$client === void 0 ? void 0 : (_this$client$disconne = _this$client.disconnect) === null || _this$client$disconne === void 0 ? void 0 : _this$client$disconne.call(_this$client);
              case 6:
                _context.next = 8;
                return (_ref3 = callbacks || _this.callbacks) === null || _ref3 === void 0 ? void 0 : (_ref3$afterDisconnect = _ref3.afterDisconnect) === null || _ref3$afterDisconnect === void 0 ? void 0 : _ref3$afterDisconnect.call(_ref3);
              case 8:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "connect", /*#__PURE__*/function () {
      var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(sessionOptions, callbacks) {
        var _ref5, _ref5$beforeConnect, _ref6, _ref6$afterConnect;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return (_ref5 = callbacks || _this.callbacks) === null || _ref5 === void 0 ? void 0 : (_ref5$beforeConnect = _ref5.beforeConnect) === null || _ref5$beforeConnect === void 0 ? void 0 : _ref5$beforeConnect.call(_ref5);
              case 2:
                if (!(_this.isMobile && _this.walletInfo.mobileDisabled)) {
                  _context2.next = 5;
                  break;
                }
                _this.setError('This wallet is not supported on mobile, please use desktop browsers.');
                return _context2.abrupt("return");
              case 5:
                _context2.prev = 5;
                _context2.t1 = _this.client;
                if (_context2.t1) {
                  _context2.next = 11;
                  break;
                }
                _context2.next = 10;
                return _this.clientPromise;
              case 10:
                _context2.t1 = _context2.sent;
              case 11:
                _context2.t0 = _context2.t1;
                if (_context2.t0) {
                  _context2.next = 16;
                  break;
                }
                _context2.next = 15;
                return _this.fetchClient();
              case 15:
                _context2.t0 = _context2.sent;
              case 16:
                _this.client = _context2.t0;
                if (_this.client) {
                  _context2.next = 20;
                  break;
                }
                _this.setClientNotExist();
                return _context2.abrupt("return");
              case 20:
                _context2.next = 22;
                return _this.update();
              case 22:
                if (sessionOptions !== null && sessionOptions !== void 0 && sessionOptions.duration) {
                  setTimeout(function () {
                    _this.disconnect(callbacks);
                  }, sessionOptions === null || sessionOptions === void 0 ? void 0 : sessionOptions.duration);
                }
                _context2.next = 28;
                break;
              case 25:
                _context2.prev = 25;
                _context2.t2 = _context2["catch"](5);
                _this.setError(_context2.t2);
              case 28:
                _context2.next = 30;
                return (_ref6 = callbacks || _this.callbacks) === null || _ref6 === void 0 ? void 0 : (_ref6$afterConnect = _ref6.afterConnect) === null || _ref6$afterConnect === void 0 ? void 0 : _ref6$afterConnect.call(_ref6);
              case 30:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, null, [[5, 25]]);
      }));
      return function (_x2, _x3) {
        return _ref4.apply(this, arguments);
      };
    }());
    _this._walletInfo = walletInfo;
    return _this;
  }
  (0, _createClass2["default"])(WalletBase, [{
    key: "walletInfo",
    get: function get() {
      return this._walletInfo;
    }
  }, {
    key: "downloadInfo",
    get: function get() {
      var _this2 = this;
      var downloads = this.walletInfo.downloads || [];
      downloads = downloads.filter(function (d) {
        var _this2$env;
        return d.device === ((_this2$env = _this2.env) === null || _this2$env === void 0 ? void 0 : _this2$env.device) || !d.device;
      });
      if (downloads.length === 1) {
        return downloads[0];
      }
      downloads = downloads.filter(function (d) {
        var _this2$env2;
        return d.os === ((_this2$env2 = _this2.env) === null || _this2$env2 === void 0 ? void 0 : _this2$env2.os) || !d.os;
      });
      if (downloads.length === 1) {
        return downloads[0];
      }
      downloads = downloads.filter(function (d) {
        var _this2$env3;
        return d.browser === ((_this2$env3 = _this2.env) === null || _this2$env3 === void 0 ? void 0 : _this2$env3.browser) || !d.browser;
      });
      return downloads[0];
    }
  }, {
    key: "walletName",
    get: function get() {
      return this.walletInfo.name;
    }
  }, {
    key: "walletPrettyName",
    get: function get() {
      return this.walletInfo.prettyName;
    }
  }, {
    key: "rejectMessageSource",
    get: function get() {
      if (typeof this.walletInfo.rejectMessage === 'string') {
        return this.walletInfo.rejectMessage;
      } else {
        var _this$walletInfo$reje;
        return (_this$walletInfo$reje = this.walletInfo.rejectMessage) === null || _this$walletInfo$reje === void 0 ? void 0 : _this$walletInfo$reje.source;
      }
    }
  }, {
    key: "rejectMessageTarget",
    get: function get() {
      if (typeof this.walletInfo.rejectMessage === 'string') {
        return void 0;
      } else {
        var _this$walletInfo$reje2;
        return (_this$walletInfo$reje2 = this.walletInfo.rejectMessage) === null || _this$walletInfo$reje2 === void 0 ? void 0 : _this$walletInfo$reje2.target;
      }
    }
  }, {
    key: "rejectCode",
    get: function get() {
      return this.walletInfo.rejectCode;
    }
  }, {
    key: "rejectMatched",
    value: function rejectMatched(e) {
      return this.rejectMessageSource && e.message === this.rejectMessageSource || this.rejectCode && e.code === this.rejectCode;
    }
  }, {
    key: "appUrl",
    get: function get() {
      return this._appUrl;
    }
  }, {
    key: "qrUrl",
    get: function get() {
      return this._qrUrl;
    }
  }, {
    key: "updateCallbacks",
    value: function updateCallbacks(callbacks) {
      this.callbacks = _objectSpread(_objectSpread({}, this.callbacks), callbacks);
    }
  }, {
    key: "setClientNotExist",
    value: function setClientNotExist() {
      this.setState(_types.State.Error);
      this.setMessage(_utils.ClientNotExistError.message);
    }
  }, {
    key: "setRejected",
    value: function setRejected() {
      this.setState(_types.State.Error);
      this.setMessage(_utils.RejectedError.message);
    }
  }, {
    key: "setError",
    value: function setError(e) {
      this.setState(_types.State.Error);
      this.setMessage(typeof e === 'string' ? e : e.message);
      if (typeof e !== 'string' && e.stack) {
        console.error(e.stack);
      }
    }
  }]);
  return WalletBase;
}(_state.StateBase);
exports.WalletBase = WalletBase;
